### PR TITLE
Add types

### PR DESCRIPTION
**Pull request by Rokt33r**
@wooorm @ChristianMurphy

### CODE REVIEW

**Comment by Rokt33r** [comment]
I disabled prettier for statements using `$ExpectError` flag. `dtslint` handles the code, formatted by prettier, differently based on version of Typescript or one of itself.

File: types/index.d.ts [hunk 1]
```diff
@@ -0,0 +1,141 @@
+// TypeScript Version: 3.0
+
+declare namespace parseEntities {
+  interface ParseEntitiesOptions<WC = any, TC = any, RC = any> {
```
Comment (ChristianMurphy) [inline types/index.d.ts]: https://devblogs.microsoft.com/typescript/announcing-typescript-3-4/#type-checking-for-globalthis

```suggestion
  interface ParseEntitiesOptions<WC = typeof globalThis, TC = typeof globalThis, RC = typeof globalThis> {
```

File: types/index.d.ts [hunk 2]
```diff
@@ -0,0 +1,141 @@
...
+  interface Location {
+    start: Position
+    end: Position
+  }
+}
+
+/**
+ * Decode special characters in `value`.
+ */
+declare function parseEntities<WC = any, TC = any, RC = any>(
```
Comment (ChristianMurphy) [inline types/index.d.ts]: https://devblogs.microsoft.com/typescript/announcing-typescript-3-4/#type-checking-for-globalthis

```suggestion
declare function parseEntities<WC = typeof globalThis, TC = typeof globalThis, RC = typeof globalThis>(
```

File: types/index.d.ts [hunk 3]
```diff
@@ -0,0 +1,141 @@
+// TypeScript Version: 3.0
```
Comment (ChristianMurphy) [inline types/index.d.ts]: https://devblogs.microsoft.com/typescript/announcing-typescript-3-4/#type-checking-for-globalthis

```suggestion
// TypeScript Version: 3.4
```

File: types/index.d.ts [hunk 4]
```diff
@@ -0,0 +1,141 @@
...
+
+  /**
+   * Error handler.
+   */
+  type ErrorHandler<C> = (
+    /**
+     * `this` refers to `warningContext` when given to `parseEntities`.
+     */
+    this: C,
+    /**
```
Comment (ChristianMurphy) [inline types/index.d.ts:71]:
```suggestion

    /**
```

File: types/index.d.ts [hunk 5]
```diff
@@ -0,0 +1,141 @@
...
+     * Human-readable reason for triggering a parse error (`string`).
+     */
+    reason: string,
+    /**
```
Comment (ChristianMurphy) [inline types/index.d.ts:76]:
```suggestion

    /**
```

File: types/index.d.ts [hunk 6]
```diff
@@ -0,0 +1,141 @@
...
+     * Place at which the parse error occurred (`Position`).
+     */
+    position: Position,
+    /**
```
Comment (ChristianMurphy) [inline types/index.d.ts:81]:
```suggestion

    /**
```

File: types/index.d.ts [hunk 7]
```diff
@@ -0,0 +1,141 @@
...
+
+  /**
+   * Text handler.
+   */
+  type TextHandler<C> = (
+    /**
+     * `this` refers to `textContext` when given to `parseEntities`.
+     */
+    this: C,
+    /**
```
Comment (ChristianMurphy) [inline types/index.d.ts:96]:
```suggestion

    /**
```

File: types/index.d.ts [hunk 8]
```diff
@@ -0,0 +1,141 @@
...
+     * String of content (`string`).
+     */
+    value: string,
+    /**
```
Comment (ChristianMurphy) [inline types/index.d.ts:101]:
```suggestion

    /**
```

File: types/index.d.ts [hunk 9]
```diff
@@ -0,0 +1,141 @@
...
+
+  /**
+   * Character reference handler.
+   */
+  type ReferenceHandler<C> = (
+    /**
+     * `this` refers to `textContext` when given to `parseEntities`.
+     */
+    this: C,
+    /**
```
Comment (ChristianMurphy) [inline types/index.d.ts:116]:
```suggestion

    /**
```

File: types/index.d.ts [hunk 10]
```diff
@@ -0,0 +1,141 @@
...
+     * String of content (`string`).
+     */
+    value: string,
+    /**
```
Comment (ChristianMurphy) [inline types/index.d.ts:121]:
```suggestion

    /**
```

File: types/index.d.ts [hunk 11]
```diff
@@ -0,0 +1,141 @@
...
+     * Location at which `value` starts and ends (`Location`).
+     */
+    location: Location,
+    /**
```
Comment (ChristianMurphy) [inline types/index.d.ts:126]:
```suggestion

    /**
```

**Review by ChristianMurphy — APPROVED** [review]

**Comment by Rokt33r** [comment]
@ChristianMurphy Fixed!

**Review by wooorm — APPROVED** [review]

**Comment by wooorm** [comment]
✨

**Comment by wooorm** [comment]
Thanks @Rokt33r! They’re both released!

**Pull request merged**